lib/_stream_writable.js

```diff
@@ -586,8 +587,17 @@ Writable.prototype.end = function(chunk, encoding, cb) {
     encoding = null;
   }
 
-  if (chunk !== null && chunk !== undefined)
-    this.write(chunk, encoding);
+  if (typeof cb === 'function') {
+    cb = once(cb);
+  }
+
+  if (chunk !== null && chunk !== undefined) {
+    this.write(chunk, encoding, cb ? function(err) {
+      if (err) {
+        cb(err);
+      }
+    } : null);
```

Comment:
if cb() is a function, this allocates 2 more closures. I think we can just limit to one by creating a “custom” `once` specifically for this.

Reply:
I'm not sure that's possible... one of them should only propagate if error while the other should always. What I could do is to move once into the if branch here... However, I believe there might be a possibility of `onFinished` invoking the callback twice in which case it would be nice to have it outside.